### ISSUE TITLE
feature: Ability to match special characters or custom patterns per keyword.

### ISSUE THREAD

**Issue by AlansCodeLog** [folke/todo-comments.nvim#213]
### Did you check the docs?

- [X] I have read all the todo-comments.nvim docs

### Is your feature request related to a problem? Please describe.

I'd like to match "???"

If I use it like so I get an error on start, if I just leave it as an alt for another keyword, the entire plugin fails silently:
```
keywords = {
	["???"] = { icon = "?", color = "default", alt = { "???" } },
}
```

Error: <details>
```
Failed to run `config` for todo-comments.nvim
vim/_editor.lua:0: VimEnter Autocommands for "*"..script nvim_exec2() called at VimEnter Autocommands for "*":0../home/alan/.local/share/nvim/sessions/%home%alan_master.vim[20]..BufReadPost Autocommands for "*"..s
cript nvim_exec2() called at BufReadPost Autocommands for "*":0: Vim(highlight):E5248: Invalid character in group name
# stacktrace:
  - vim/_editor.lua:0 _in_ **cmd**
  - /todo-comments.nvim/lua/todo-comments/config.lua:187 _in_ **colors**
  - /todo-comments.nvim/lua/todo-comments/config.lua:133 _in_ **_setup**
  - /todo-comments.nvim/lua/todo-comments/config.lua:95 _in_ **setup**
  - vim/_editor.lua:0
  - /auto-session/lua/auto-session/init.lua:658 _in_ **restore**
  - /auto-session/lua/auto-session/init.lua:720 _in_ **AutoRestoreSession**
  - /auto-session/lua/auto-session/init.lua:908
  ```
</details>

I also tried escaping the question marks like `\\?` and `%?`, but that also didn't work, though I could be doing something wrong, I'm not that familiar with lua.

### Describe the solution you'd like

Auto escaping of special characters or I'm comfortable just specifying a custom pattern for that keyword.

### Describe alternatives you've considered

.

**Comment by KeithETruesdell** [comment]
I thought i was the only crazy person that did this!  3 or 4 question marks.  
I tried to fiddle with this as well, and even had some custom patterns to match and nothing.
I noticed that if you have special characters in the keywords, especially the ? It will break.  I tried quoting it, and i also tried having it as a separate item in the "alt" variable for a legit keyword ( e.g. - HUH = { ... alt = { "???"}} And that didnt work.
I tried a few different patterns that normally work in regex but didnt work here.
If you got it to work, let me know, otherwise i am going to try to solve this.

**Comment by caliguIa** [comment]
I also am having this issue but with the '@' character, would love a solution. My colleagues use PHPStorm, who's todo comments are '@todo'. Did anyone find a work around for this?

**Comment by dsaad68** [comment]
I have the same issue:

![Screenshot 2024-06-09 010109](https://github.com/folke/todo-comments.nvim/assets/6844325/76762056-54e6-4bc9-a285-101a403e0866)

I'm using `nvchad`, and this is my config: 

```lua
require("todo-comments").setup {
  signs = true, -- show icons in the signs column
  sign_priority = 8, -- sign priority
  -- keywords recognized as todo comments
  keywords = {
    FIX = {
      icon = "", -- icon used for the sign, and in search results
      color = "fix", -- can be a hex color, or a named color (see below)
      alt = { "FIXME", "BUG", "FIXIT", "ISSUE" }, -- a set of other keywords that all map to this FIX keywords
    },
    TODO = { icon = "✅", color = "todo" },
    INFO = { icon = "ℹ️ ", color = "info", alt = { "info", "INFO" } },
    IDEA = { icon = "💡", color = "idea" },
    NOTE = { icon = "📓", color = "note" },
    STEP = { icon = "🪜", color = "step" },
    TASK = { icon = "✔️ ", color = "task", alt = { "[ ]", "%[ %]", "%[%s%]" } },
    DONE = { icon = "☑️ ", color = "done", alt = { "[x]", "%[x%]" } },
    LEARN = { icon = "📖", color = "learn" },
    CHECK = { icon = "⚠️", color = "check" },
    TEST = { icon = "🧪", color = "task" },
  },

  -- list of named colors where we try to extract the guifg from the
  -- list of highlight groups or use the hex color if hl not found as a fallback
  colors = {
    fix = { "#FF2D00" },
    todo = { "#FF8C00" },
    info = { "#0C7C59" },
    check = { "#CA686B" },
    learn = { "#0071B1" },
    step = { "#92524f" },
    task = { "#c345e9" },
    done = { "#521d63" },
    idea = { "#861657" },
    note = { "#569cd6" },
  },

  gui_style = {
    fg = "NONE", -- The gui style to use for the fg highlight group.
    bg = "BOLD", -- The gui style to use for the bg highlight group.
  },
  merge_keywords = true, -- when true, custom keywords will be merged with the defaults

  -- highlighting of the line containing the todo comment
  -- * before: highlights before the keyword (typically comment characters)
  -- * keyword: highlights of the keyword
  -- * after: highlights after the keyword (todo text)
  highlight = {
    multiline = true, -- enable multine todo comments
    multiline_pattern = "^.", -- lua pattern to match the next multiline from the start of the matched keyword
    multiline_context = 10, -- extra lines that will be re-evaluated when changing a line
    before = "fg", -- "fg" or "bg" or empty
    keyword = "fg", -- "fg", "bg", "wide", "wide_bg", "wide_fg" or empty. (wide and wide_bg is the same as bg, but will also highlight surrounding characters, wide_fg acts accordingly but with fg)
    after = "fg", -- "fg" or "bg" or empty
    pattern = [[.*<(KEYWORDS)\s*:]], -- pattern or table of patterns, used for highlighting (vim regex)
    comments_only = true, -- uses treesitter to match keywords in comments only
    max_line_len = 400, -- ignore lines longer than this
    exclude = {}, -- list of file types to exclude highlighting
  },

  search = {
    command = "rg",
    args = {
      "--color=never",
      "--no-heading",
      "--with-filename",
      "--line-number",
      "--column",
    },

    -- regex that will be used to match keywords.
    -- don't replace the (KEYWORDS) placeholder
    pattern = [[\b(KEYWORDS):]], -- ripgrep regex
    -- pattern = [[\b(KEYWORDS)\b]], -- match without the extra colon. You'll likely get false positives
  },
}
```

**Comment by b0ae989c** [comment]
There is no workaround right now, because highlights depend on keyword names. Those names have to be a simple string without special characters. See

https://github.com/folke/todo-comments.nvim/blob/70a93ce66083699571adc361166504b03cc39c2b/lua/todo-comments/config.lua#L187-L189

To support this feature, we need to parse keyword names and replace every special character with a valid string. For example,
- `?` => `QuestionMark`
- `[` => `LeftBracket`
- `%` => `Percentage`

**Comment by AlansCodeLog** [comment]
Please don't close.

**Comment by HenrysHouses** [comment]
i ran into this too just now, i have comments all over the codebase with ?, !, and #. as keywords. as said symbols doesnt work, but putting ? in there breaks the entire plugin without any warning or errors

**Comment by soulshined** [comment]
Just adding my 2 cents that this would be a nice to have. I personally have a lot of git hooks that look for annotations, so for example:

```
@CRITICAL update xyz before push -- @NOCHECKIN 
func foobar()
```

the git hook acts accordingly but it would be nice to have it highlighted. I find that adding symbols is a better way to visualize keywords since there are scenarios where you would type the keywords literal value in the comment itself, `fix` is a good example I think, but that's also not fair to generalize since I don't have the colon suffix requirement in my search pattern. Just food for thought though. 

Appreciate all the work you've done for plugins btw. Coming from vscode so it's been an easy transition thanks to all your work
